src/sa.py

```diff
@@ -1,15 +1,277 @@
+        for element in bucket_out[key]:
+            if element in LMS_array:
+                block = LMS_blocks[LMS_array.index(element)]
+                if block != previous:
+                    count += 1
+                previous = block
+                reduced[LMS_array.index(element)] = str(count)
+    return "".join(reduced)
+
+def compute_t1(string):
+    LS_array = annotate_string(string)
+    LMS_array = get_LMS(LS_array)
+    LMS_blocks = get_LMS_blocks(string, LMS_array)
+    buckets = impute_buckets(string)
+    bucket_out = output_bucket(buckets)
+    bucket_out = first_pass(buckets, bucket_out, LMS_array, LMS_blocks)
+    bucket_out = second_pass(bucket_out, LS_array, string)
+    bucket_out = third_pass(bucket_out, LS_array, string)
+    reduced = 0#number_LMS(bucket_out, LMS_blocks, LMS_array)
+    return reduced, bucket_out
+
+#def SAIS(string):
+#    stack = list()
+#    while True:
+#        reduced, bucket_out = compute_t1
+#        if reduced[-1] != len()
+
+#def mostly_naive(string):
+#    forget, bucket_out = compute_t1(string)
+#    SA = []
+#    for i in bucket_out:
+#        SA.extend(bucket_out[i])
+#    SA = [string[SA[int(i)]:] for i in SA]
+#    SA.sort()
+#    return(SA)
+
+def completely_naive(string):
+    SA = [i for i in range(len(string))]
+    SA = [string[SA[int(i)]:] for i in SA]
+    SA.sort()
+    return(SA)
+
+def test_SA(SA):
+    for i in SA:
+        print(i)
+
+def LCP_naive(SA, string):
+    LCP = [0, 0]
+    for i in range(2, len(SA)):
+        count = 0
+        for j in range(len(string)):
+            if string[SA[i] + j] == string[SA[i - 1] + j]:
+                count += 1
+            else:
+                break
+        LCP.append(count)
+    return LCP
+
+def naive(string):
+    string += "$"
+    SA = completely_naive(string)
+    SA = [len(string) - len(i) for i in SA]
+    LCP = LCP_naive(SA, string)
+    return SA, LCP
+
+def compare_strings(pattern, suffix):
+    loops = min(len(pattern), len(suffix))
+
+    for i in range(loops):
+        if pattern[i] < suffix[i]:
+            return False
+        if pattern[i] > suffix[i]:
+            return True
+
+def SA_search(SA, string, pattern):
+    lo = 0
+    hi = len(SA) - 1
+    while True:
+        mid = int((lo + hi)/2)
+        suffix = string[SA[mid]:]
+        match = compare_strings(pattern, suffix)
+        if match is None:
+            return mid
+        elif hi <= lo:
+            return None
+        elif match:
+            lo = mid + 1
+        elif not match:
+            hi = mid - 1
+
+def SA_match(SA, LCP, string, pattern):
+    string += "$"
+    match = SA_search(SA, string, pattern)
+    l = len(pattern)
+    if match is None:
+        return []
+    lo = match
+    hi = match + 1
+
+    while hi < len(LCP):
+        if LCP[hi] >= l:
+            hi += 1
+        else:
+            break
+    while lo > 0:
+        if LCP[lo] >= l:
+            lo -= 1
+        else:
+            break
+    return SA[lo:hi]

```

Comment:
Ok, yes that is a fine use of `LCP`. (You can also use it to speed up the actual binary search, but there it is a little trickier).